pkg/network/usm/sharedlibraries/ebpf.go

```diff
@@ -246,73 +248,110 @@ func (e *EbpfProgram) InitWithLibsets(libsets ...Libset) error {
 
 	e.setupManagerAndPerfHandlers()
 
-	if err := e.loadProgram(libsets); err != nil {
+	// Mark the libsets as requested so they can be started
+	// Note that other libsets might be requested from previous executions
+	for _, libset := range libsets {
+		e.libsets[libset].requested = true
+	}
+
+	if err := e.loadProgram(); err != nil {
 		return fmt.Errorf("cannot load program: %w", err)
 	}
 
 	if err := e.start(); err != nil {
 		return fmt.Errorf("cannot start manager: %w", err)
 	}
 
-	// Add the libsets to the enabled libsets
-	for _, libset := range libsets {
-		e.enableLibset(libset)
-	}
 	e.isInitialized = true
 	return nil
 }
 
+// start starts the eBPF program and the perf handlers, assumes the initMutex is locked
 func (e *EbpfProgram) start() error {
 	err := e.Manager.Start()
 	if err != nil {
 		return fmt.Errorf("cannot start manager: %w", err)
 	}
 
-	for libset, data := range e.libsets {
-		if !data.enabled {
+	for _, handler := range e.libsets {
+		if !handler.requested {
 			continue
 		}
 
+		// Init the "done" channel for the handler, it will be closed when the handler stops
+		handler.done = make(chan struct{})
 		e.wg.Add(1)
-		go func(ls Libset, ldata *libsetData) {
-			defer e.wg.Done()
-
-			dataChan := ldata.perfHandler.DataChannel()
-			lostChan := ldata.perfHandler.LostChannel()
-			for {
-				select {
-				case <-data.done:
-					return
-				case event, ok := <-dataChan:
-					if !ok {
-						return
-					}
-					e.handleEvent(&event, ls)
-				case <-lostChan:
-					// Nothing to do in this case
-					break
-				}
-			}
-		}(libset, data)
+		go handler.eventLoop(&e.wg)
+
+		handler.enabled = true
 	}
 
 	return nil
 }
 
-func (e *EbpfProgram) handleEvent(event *ebpf.DataEvent, libset Libset) {
-	e.callbacksMutex.RLock()
-	defer func() {
-		event.Done()
-		e.callbacksMutex.RUnlock()
-	}()
+// eventLoop is the main loop for a single libset. Should be called with all perfHandlers initialized.
+func (l *libsetHandler) eventLoop(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	dataChan := l.perfHandler.DataChannel()
+	lostChan := l.perfHandler.LostChannel()
+	for {
+		select {
+		case <-l.done:
+			return
+		case event, ok := <-dataChan:
+			if !ok {
+				return
+			}
+			l.handleEvent(&event)
+		case <-lostChan:
+			// Nothing to do in this case
+			break
+		}
+	}
+}
+
+func (l *libsetHandler) handleEvent(event *ebpf.DataEvent) {
+	defer event.Done()
+	l.callbacksMutex.RLock()
+	defer l.callbacksMutex.RUnlock()
 
 	libpath := ToLibPath(event.Data)
```

Comment:
nit -> I'd move `libpath := ToLibPath(event.Data)` to be before the mutex lock, as it is not related to the callbacks